ota_proxy/db.py

```diff
@@ -55,91 +88,54 @@ def close(self):
 
     def _init_table(self):
         logger.debug("init sqlite database...")
-        cur = self._con.cursor()
-
-        # create the table
-        cur.execute(
-            f"""CREATE TABLE {self.TABLE_NAME}(
-                    url text UNIQUE PRIMARY KEY, 
-                    hash text NOT NULL, 
-                    size real NOT NULL,
-                    content_type text, 
-                    content_encoding text)"""
-        )
-
-        self._con.commit()
-        cur.close()
+        with self._general_query(self.INIT_DB, (), init=True):
+            self._con.commit()
 
     def _connect_db(self, init: bool):
         if init:
-            # remove the old db file if needed
             Path(self._db_file).unlink(missing_ok=True)
 
         self._con = sqlite3.connect(self._db_file, check_same_thread=False)
+        self._con.row_factory = sqlite3.Row
 
         # check if the table exists
-        cur = self._con.cursor()
-        cur.execute(
+        with self._general_query(
             "SELECT name FROM sqlite_master WHERE type='table' AND name=?",
             (self.TABLE_NAME,),
-        )
-        if cur.fetchone() is None:
-            self._init_table()
+            init=True,
+        ) as cur:
+            if cur.fetchone() is None:
+                self._init_table()
 
     def remove_url_by_hash(self, *hash: str):
-        if self._closed:
-            raise sqlite3.OperationalError("connect is closed")
-
-        hash_list = [(h,) for h in hash]
-        with self._wlock:
-            cur = self._con.cursor()
-            cur.executemany(f"DELETE FROM {self.TABLE_NAME} WHERE hash=?", hash_list)
-
+        hash = [(h,) for h in hash]
+        with self._wlock, self._general_query(
+            f"DELETE FROM {self.TABLE_NAME} WHERE hash=?", hash
+        ):
             self._con.commit()
-            cur.close()
-        return
 
     def remove_urls(self, *urls: str):
-        if self._closed:
-            raise sqlite3.OperationalError("connect is closed")
-
-        url_list = [(u,) for u in urls]
-        with self._wlock:
-            cur = self._con.cursor()
-            cur.executemany(f"DELETE FROM {self.TABLE_NAME} WHERE url=?", url_list)
-
+        urls = [(u,) for u in urls]
+        with self._wlock, self._general_query(
+            f"DELETE FROM {self.TABLE_NAME} WHERE url=?", urls
```

Comment:
Question: Can we make sure that urls don't contain `;` string? I'd like to make sure that query sanitizing is required or not.

Reply:
using `?` mark with input value is the standard way to assemble query string according to the python3 sqlite3 document. With `?` mark, I think the sqlite3 module will process the input value properly.
``` python3
# copied from python3 sqlite3 document

# Never do this -- insecure!
symbol = 'RHAT'
cur.execute("SELECT * FROM stocks WHERE symbol = '%s'" % symbol)

# Do this instead
t = ('RHAT',)
cur.execute('SELECT * FROM stocks WHERE symbol=?', t)
print(cur.fetchone())

# Larger example that inserts many records at a time
purchases = [('2006-03-28', 'BUY', 'IBM', 1000, 45.00),
             ('2006-04-05', 'BUY', 'MSFT', 1000, 72.00),
             ('2006-04-06', 'SELL', 'IBM', 500, 53.00),
            ]
cur.executemany('INSERT INTO stocks VALUES (?,?,?,?,?)', purchases)
```